activesupport/lib/active_support/notifications/event_assertions.rb

```diff
@@ -0,0 +1,87 @@
+# typed: true
+# frozen_string_literal: true
+
+module ActiveSupport
+  module Notifications
+    module EventAssertions
+      # Assert an event was emitted with the given +name+ and +payload+.
+      #
+      # You can assert that an event was emitted by passing a name, an optional
+      # payload, and a block. While the block is executed, if a matching event
+      # is emitted, the assertion will pass. Otherwise, it will fail.
+      #
+      #     assert_event("post.submitted", title: "Cool Post") do
+      #       post.submit(title: "Cool Post") # => emits matching ActiveSupport::Notifications::Event
+      #     end
+      #
+      def assert_event(name, payload = nil, &block)
+        events = []
+
+        ActiveSupport::Notifications.subscribe(name) do |*args|
+          events << ActiveSupport::Notifications::Event.new(*args)
+        end
+
+        yield
+
+        selected_events = events.select { |event| event.name == name }
+        assert_not_empty(selected_events, "No #{name} events were found.")
+
+        return if payload.nil?
+
+        event = selected_events.find { |event| event.payload == payload }
+        assert_not_nil(
+          event,
+          "No #{name} event with payload #{payload} was found. " \
+            "Did you mean one of these payloads: #{selected_events.map(&:payload).join("\n")}?",
+        )
+      end
```

Comment:
Other test helpers in this style (like [ActionMailer::TestHelper](https://edgeapi.rubyonrails.org/classes/ActionMailer/TestHelper.html#method-i-capture_emails) and [ActionCable::TestHelper](https://edgeapi.rubyonrails.org/classes/ActionCable/TestHelper.html#method-i-capture_broadcasts)) expose a `capture_`-prefixed helper that the assertions also invoke. 

Have you explored what that might involve for events? Maybe something like:

```suggestion
      def assert_event(name, payload = nil, &block)
        events = capture_events(name, &block)
        selected_events = events.select { |event| event.name == name }
        assert_not_empty(selected_events, "No #{name} events were found.")

        return if payload.nil?

        event = selected_events.find { |event| event.payload == payload }
        assert_not_nil(
          event,
          "No #{name} event with payload #{payload} was found. " \
            "Did you mean one of these payloads: #{selected_events.map(&:payload).join("\n")}?",
        )
      end
      
      def capture_events(name, &block)
        events = []

        ActiveSupport::Notifications.subscribe(name) do |*args|
          events << ActiveSupport::Notifications::Event.new(*args)
        end

        yield
        
        events
      end
```

Reply:
Had not noticed that pattern before. Will be quite nice to de-duplicate the event subscribing and yielding pattern with a simple, reusable `capture_events` private method. Will push a cleanup.